b2g/chrome/content/remote_control.js

```diff
@@ -20,57 +20,99 @@ const kPrefPresentationDiscoverable = "dom.presentation.discoverable";
+      debug("observe: " + topic + ", " + data);
 
-  var networkStatusMonitor = {
-    // nsIObserver
-    observe: function(subject, topic, data) {
       switch (topic) {
         case "network-active-changed": {
           if (!subject) {
             // Stop service when there is no active network
-            remoteControlScope.RemoteControlService.stop();
+            this.stopService();
             break;
           }
 
           // Start service when active network change with new IP address
           // Other case will be handled by "network:offline-status-changed"
-          if (!Services.io.offline && Services.prefs.getBoolPref(kPrefPresentationDiscoverable)) {
-            importRemoteControlService();
-            remoteControlScope.RemoteControlService.start();
+          if (!Services.io.offline) {
+            this.startService();
           }
+
           break;
         }
         case "network:offline-status-changed": {
           if (data == "offline") {
             // Stop service when network status change to offline
-            remoteControlScope.RemoteControlService.stop();
-          } else if (Services.prefs.getBoolPref(kPrefPresentationDiscoverable)) {
+            this.stopService();
+          } else { // online
             // Resume service when network status change to online
-            importRemoteControlService();
-            remoteControlScope.RemoteControlService.start();
+            this.startService();
           }
           break;
         }
+        case "nsPref:changed": {
+          if (data != kPrefPresentationDiscoverable) {
+            break;
+          }
+
+          if (!this.startService()) {
+            this.stopService();
+          }
+
+          break;
+        }
         default:
           break;
       }
-    }
-  };
+    },
 
-  if (Ci.nsINetworkManager) {
-    Services.obs.addObserver(networkStatusMonitor, "network-active-changed", false);
-    Services.obs.addObserver(networkStatusMonitor, "network:offline-status-changed", false);
-  }
+    onTimeChange: function sm_onTimeChange() {
+      debug("onmoztimechange");
+      // Notice remote-control service the time is changed
+      remoteControlScope.RemoteControlService &&
+      remoteControlScope.RemoteControlService.onTimeChange(getTime());
+    },
+
+    startService: function sm_startService() {
+      debug("startService");
+      // Start only when presentation is discoverable
+      if (this.discoverable) {
+        importRemoteControlService();
+        remoteControlScope.RemoteControlService.start();
+        return true;
+      }
+      return false;
+    },
+
+    stopService: function sm_stopService() {
+      debug("stopService");
+      remoteControlScope.RemoteControlService &&
+      remoteControlScope.RemoteControlService.stop();
+    },
+
+    get discoverable() {
+      return Services.prefs.getBoolPref(kPrefPresentationDiscoverable);
+    },
+  };
 
-  setupRemoteControl();
+  statusMonitor.init();
```

Comment:
I mentioned this issue on [0]. If remote-control service is started on boot-up, and it's restarted on receiving time-change event, then we will get the following sequence:

1. flash
  1. Init time is 2016/xx/xx
  2. start service -> getOrCreateCert -> Get cert fail -> create cert 'A' started from 2016... 
  3. onTimeChange -> time is 2016/xx/xx -> getOrCreateCert -> Get cert 'A' -> validate successfully
2. first boot
  1. Init time is 2000/xx/xx
  2. start service -> getOrCreateCert -> Get cert 'A' -> validate fail -> create cert 'B' started from 2000...
  3. onTimeChange -> time is 2016/xx/xx -> getOrCreateCert -> Get cert 'B' -> validate fail -> create cert 'B_1' started from 2016...
3. second boot
  1. Init time is 2000/xx/xx
  2. start service -> getOrCreateCert -> Get cert 'B_1' -> validate fail -> create cert 'C' started from 2000...
  3. onTimeChange -> time is 2016/xx/xx -> getOrCreateCert  -> Get cert 'C' -> validate fail -> create cert 'C_1' started from 2016...
...
...
...
4. .....

When |getOrCreateCert|[1] is called for starting service, it will validate the certificate's period. If current time < cert's started day or (current time - cert's expired day) < 1 day, then it's invalid[2]. 

Thus, if the code here is
```javascript
statusMonitor.init();

if (statusMonitor.startService()) {
  statusMonitor.stopService();
}
```
The certificate will be changed all the times.

[0] https://bugzilla.mozilla.org/show_bug.cgi?id=1278767#c24
[1] https://dxr.mozilla.org/mozilla-central/source/security/manager/ssl/nsILocalCertService.idl#30
[2] http://searchfox.org/mozilla-central/source/security/manager/ssl/LocalCertService.cpp#331